qip-0009.md

```diff
@@ -0,0 +1,529 @@
+ ```
+  QIP: 9
+  Layer: Consensus (hard fork)
+  Title: Interlinks in Quai
+  Author: gameofpointers <gameofpointers+qip@quai.org>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/quainetwork/qips/wiki/Comments:QIP-0009
+  Status: Draft
+  Type: Standards Track
+  Created: 2023-11-16
+  License: BSD-2-Clause
+ ```
+
+## Abstract
+
+This QIP defines a block interlink structure to be used in Quai as described in
+[1] with few modifications.
+
+## Motivation
+
+To Prove/Verify the Proof Of Work done on Quai since the genesis using a
+non-interactive proof of proof of work (NiPoPoW). Also verify a transaction inclusion in a Block at a depth for which the light node(SPV client) doesn't have the state to verify. This requires using only
+O(log(N)) headers instead of the whole chain history. Specifically in Quai, the plan is to use interlinks for work based
+snap sync and to use nodes as trustless RPC providers.
+
+## Specification
+
+### Definitions
+1. $Order(O)$ - Quai blockchain is merged mined hierarchy of blockchains organized into different levels, Prime, Region and Zone. Prime has a order of 0, Region has order of 1, Zone has order 2. Order of a Block(B) is defined as $O(B)$. $O(B_p)=0$, $O(B_r)=1$, and $O(B_z)=2$. The organization of the levels and merged mining is explained in the [qip0005](qip-0005.md) spec.
+
+### Overview
+
+Interlink data structure is a skip-list that helps a verifier to process a
+sparse subset of the blockchain.
+
+Proof of Work function of a Valid Block in the blockchain satisfies the Mining Target(T) i.e $id \leq T$ . 
+
+Traditional Interlinks are organized into different levels. If $id \leq
+\frac{T}{2^{k \times \mu}}$ we say the block is of rank $\mu$. k is a compression
+parameter and determines the range between interlinks. This is a small
+deviation from the equation in [1] $id \leq \frac{T}{2^{\mu}}$. Having a bigger
+range between interlink levels helps with proof compression while preserving
+the security guarantees.Rank in Quai is analogous to Level in paper[1]. Since Quai has three layers namely Prime, Region, Zone. Prime chain acts as the base chain and has rank of 0. Region chain has a rank of -1 and Zone chain has a rank of -2 by definition. [1] defines $\mu$ for a block $B$ as $$\mu=\log(T)-\log(id(B))$$
+This changes in Quai to $$\mu=\frac{\log(T)-\log(id(B)}{k} $$ For blocks with rank higher than Prime ($\mu=0$).
+
+Genesis Block is set to have $id=0$ and $\mu=\inf$.
+
+Now for the implemention of interlinks in Quai it's established that,
+* Frequency of interlink computation will be on every Prime Block
```

Comment:
Howso? The node has to build the interlink into the block header before he knows if the block will be a prime/region/zone block.

And since the interlink structure needs to account for coordinate prime blocks, it could change many times in a zone or region before that zone/region finds a prime block.

Regardless.... these are all implementation details, which I don't think belong in this document

Reply:
okay, will remove this reference to frequency

Reply:
`And since the interlink structure needs to account for coordinate prime blocks, it could change many times in a zone or region before that zone/region finds a prime block.` -> This is what i meant with the frequency of the prime block